## Pandas 실습 예제 - Titanic Dataset


In [1]:
from IPython.display import Image
Image(url= "https://static1.squarespace.com/static/5006453fe4b09ef2252ba068/5095eabce4b06cb305058603/5095eabce4b02d37bef4c24c/1352002236895/100_anniversary_titanic_sinking_by_esai8mellows-d4xbme8.jpg", width=600, height=300)

[RMS 타이타닉: Wikipedia](https://ko.wikipedia.org/wiki/RMS_%ED%83%80%EC%9D%B4%ED%83%80%EB%8B%89)




In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

## Colab 파일 업로드

In [4]:
# PC에 저장된 파일 업로드
from google.colab import files
myFile = files.upload()

# content 폴더에 파일 업로드
# 단, 코랩 접속을 종료하면 content 초기화
# 매번 파일 업로드 작업 진행 필요

Saving titanic_train.csv to titanic_train.csv


## Data Frame
- CSV 파일 불러오기

In [48]:
# titanic_train.csv 파일을 데이터프레임으로 저장
titanic_df = pd.read_csv('titanic_train.csv')

# content 폴더를 현재 위치로 인식하므로 경로 지정 불필요

In [6]:
# titanic_df 상위 5개 데이터 조회
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### 컬럼 정보
- Passengerid: 탑승자 일련번호
- Survived: 생존여부(0: 사망, 1: 생존)
- Pclass: 선실 등급(1: 일등석, 2: 이등석, 3: 삼등석)
- Sex: 탑승자의 성별
- Name: 탑승자 이름
- Age: 탑승자 나이
- Sibsp: 같이 탑승한 형제자매 또는 배우자 인원수
- Parch: 같이 탑승한 부모님 또는 어린이 인원수
- Ticket: 티켓 번호
- Fare: 요금
- Cabin: 선실 번호
- Embarked: 탑승 항구(C: Cherbourg, Q: Queenstown, S: Southampton)


In [7]:
# titanic_df 데이터 타입 확인
type(titanic_df)

pandas.core.frame.DataFrame

In [8]:
# 행과 열의 크기 확인
titanic_df.shape

# 튜플 형태로 shape 반환 -> 인덱싱 가능

(891, 12)

In [9]:
# 컬럼명 정보 확인
titanic_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [10]:
# 컬럼명, 데이터 타입, 결측치(null) 개수 조회
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [11]:
# 수치형 데이터의 기술통계량(평균값, 최댓값/최솟값, n-percentile 분포 등) 확인
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [12]:
# 문자형 데이터의 통계량 확인(include=['O'] -> include object)
titanic_df.describe(include=['O'])

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

- count : 해당 열에서 NaN을 제외한 데이터의 개수

- unique : 해당 열에서 중복되지 않는 값의 개수

- top : 해당 열에서 가장 많이 중복되는 데이터

- freq : 가장 많이 중복되는 데이터의 개수

- 전체 891명 중 577명이 남성

- 210개의 중복된 티켓 번호(ticket) 존재

- 선실 번호(cabin)에 중복값이 있는 것으로 보아 같은 방을 사용한 인원 존재

In [13]:
# Pclass 컬럼 데이터 분포 확인
titanic_df['Pclass'].value_counts()

# unique 데이터별 개수
# value_counts()는 Series를 대상으로 사용하는 함수

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [14]:
# Pclass 컬럼(Series)의 값을 titanic_pclass 변수에 저장
titanic_pclass = titanic_df['Pclass']
type(titanic_pclass)

pandas.core.series.Series

In [15]:
# 상위 5개 데이터 조회
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [16]:
# Pclass 데이터 종류별 개수를 변수에 저장
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
type(value_counts) # value.counts(): Series 반환

3    491
1    216
2    184
Name: Pclass, dtype: int64


pandas.core.series.Series

In [17]:
# 전체 인원, 생존/사망 인원
print(f"전체 인원: {titanic_df.shape[0]}")
print(f"생존자 수: {titanic_df['Survived'].value_counts()[1]}") # 인덱스 1: 생존자
print(f"사망자 수: {titanic_df['Survived'].value_counts()[0]}") # 인덱스 0: 사망자

전체 인원: 891
생존자 수: 342
사망자 수: 549


In [18]:
# 남성/여성 인원
print(f"남성 수: {titanic_df['Sex'].value_counts()['male']}")
print(f"여성 수: {titanic_df['Sex'].value_counts()['female']}")

남성 수: 577
여성 수: 314


## 컬럼 추출

In [22]:
# Pclass 컬럼 상위 5개 데이터 조회
titanic_df[['Pclass']].head()

# 이중 대괄호 사용 시 dataframe 형태로 출력 가능

Pclass
0       3
1       1
2       3
3       1
4       3

In [24]:
# 여러 컬럼의 데이터 추출(list 사용)
titanic_df[['Pclass', 'Survived']].head()

Pclass  Survived
0       3         0
1       1         1
2       3         1
3       1         1
4       3         0

In [30]:
# 상위 2개의 행을 슬라이싱하여 조회
titanic_df[0:2]

# 단, titanic_df[0] 형태의 단일 인덱싱 불가능 -> KeyError 발생

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

## 컬럼 생성 및 수정

In [31]:
# Age_0 컬럼 생성(모든 데이터를 0으로 설정)
titanic_df['Age_0'] = 0
titanic_df['Age_0'].head()

0    0
1    0
2    0
3    0
4    0
Name: Age_0, dtype: int64

In [32]:
# Age_by_10 컬럼 생성: Age 컬럼 데이터 * 10 
titanic_df['Age_by_10'] = titanic_df['Age'] * 10
titanic_df['Age_by_10'].head()

0    220.0
1    380.0
2    260.0
3    350.0
4    350.0
Name: Age_by_10, dtype: float64

In [33]:
# Age_by_10 컬럼 수정: 기존 값 + 100
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df['Age_by_10'].head()

0    320.0
1    480.0
2    360.0
3    450.0
4    450.0
Name: Age_by_10, dtype: float64

In [34]:
# Family_no 컬럼 생성: SibSp + Parch
titanic_df['Family_no'] = titanic_df['SibSp'] + titanic_df['Parch']
titanic_df['Family_no'].head()

0    1
1    1
2    0
3    1
4    0
Name: Family_no, dtype: int64

In [35]:
# Family_no 컬럼 수정: 기존 값 + 1
titanic_df['Family_no'] = titanic_df['Family_no'] + 1
titanic_df['Family_no'].head()

0    2
1    2
2    1
3    2
4    1
Name: Family_no, dtype: int64

## 데이터 삭제
- DataFrame.drop(labels=None, axis=0, inplace=False)
- column 삭제
 - labels='column_name', axis=1
 - 기존 컬럼을 가공한 새로운 컬럼 생성 후 기존 컬럼 제거
- row 삭제
 - labels='row_index', axis=0
 - 이상치 제거

### 컬럼 삭제

In [36]:
# Age_0 컬럼을 삭제한 데이터 조회(원본 데이터 유지)
titanic_df.drop('Age_0', axis=1)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_no  
0        0         A/5 21171   7.2500   NaN        S      320.0          2  
1        0          PC 17599  71.2833   C85        C      480.0          2  
2        0  STON/O2. 3101282   7.9250   NaN        S      360.0          1  
3        0            113803  53.1000  C123        S      450.0          2  
4        0            373450   8.0500   NaN        S      450.0          1  
..     ...               ...      ...   ...      ...        ...        ...  
886      0            211536  13.0000   NaN        S      370.0          1  
887      0            112053  30.0000   B42        S      290.0          1  
888      2        W./C. 6607  23.4500   NaN        S        NaN          4  
889      0            111369  30.0000  C148        C      360.0          1  
890      0            370376   7.7500   NaN        Q      420.0          1  

[891 rows x 14 columns]

In [39]:
# 컬럼 삭제(원본 데이터 수정)

# 1) inplace=True
# titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)

# 2) 컬럼이 삭제된 데이터를 변수에 다시 저장
# titanic_df = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1)

# 단, 두 가지 방식 동시에 적용 불가능
# inplace=True 옵션을 사용 시 반환값: None -> 변수에 아무것도 저장 x

In [38]:
# Age_0, Age_by_10, Family_No 컬럼 삭제
titanic_df = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_no'], axis=1)
titanic_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### 행 삭제

In [41]:
# 원본 데이터 조회(행 삭제 이전)
print('***** before drop *****')
print()
titanic_df.head()

***** before drop *****



PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [49]:
# 상위 3개의 행이 삭제된 데이터 조회(원본 데이터 유지)
print('***** after drop *****')
print()
titanic_df.drop([0, 1, 2]).head()

***** after drop *****



PassengerId  Survived  Pclass  \
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   
5                              Moran, Mr. James    male   NaN      0      0   
6                       McCarthy, Mr. Timothy J    male  54.0      0      0   
7                Palsson, Master. Gosta Leonard    male   2.0      3      1   

   Ticket     Fare Cabin Embarked  
3  113803  53.1000  C123        S  
4  373450   8.0500   NaN        S  
5  330877   8.4583   NaN        Q  
6   17463  51.8625   E46        S  
7  349909  21.0750   NaN        S

## 위치 기반 데이터 인덱싱
- iloc()


In [50]:
# titanic_df 상위 5개 데이터 조회(iloc 사용)
titanic_df.iloc[0:5]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [51]:
# 상위 5개 데이터의 두 번째 컬럼 조회
titanic_df.iloc[0:5, 1]

0    0
1    1
2    1
3    1
4    0
Name: Survived, dtype: int64

In [52]:
# 상위 5개 데이터의 1, 3, 5번 인덱스 컬럼 조회(2, 4, 6번째 컬럼)
titanic_df.iloc[0:5, [1, 3, 5]]

Survived                                               Name   Age
0         0                            Braund, Mr. Owen Harris  22.0
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0
2         1                             Heikkinen, Miss. Laina  26.0
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0
4         0                           Allen, Mr. William Henry  35.0

## 명칭 기반 데이터 인덱싱
- loc()
- 행 위치에는 인덱스 값, 열 위치에는 컬럼명 입력


In [53]:
# Age 컬럼의 1~10번 인덱스 데이터 조회
titanic_df.loc[1:10, 'Age']

1     38.0
2     26.0
3     35.0
4     35.0
5      NaN
6     54.0
7      2.0
8     27.0
9     14.0
10     4.0
Name: Age, dtype: float64

In [57]:
# 'Heikkinen, Miss. Laina' 출력(loc 사용)
print(titanic_df.loc[2, 'Name'])
print()
titanic_df.head()

Heikkinen, Miss. Laina



PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [65]:
# loc(): 슬라이싱 범위에 end 값 포함
print('iloc 인덱싱')
print(titanic_df.iloc[0:1, 3]) # end 값 미포함
print()
print('loc 인덱싱')
print(titanic_df.loc[0:1, 'Name']) # end 값 포함

iloc 인덱싱
0    Braund, Mr. Owen Harris
Name: Name, dtype: object

loc 인덱싱
0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
Name: Name, dtype: object


## 조건 기반 데이터 인덱싱
- 불린(boolean) 인덱싱 : 조건을 이용하여 원하는 데이터를 조회하는 방식
- iloc, loc처럼 인덱스를 지정하는 방식보다 조건을 지정하는 방식을 자주 사용
- 복합 조건 : &(and), |(or), ~(not) 

In [66]:
# 65세 이상인지 여부를 조회(True/False)
titanic_df['Age'] >= 65

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [67]:
# 65세 이상인 승객의 데이터 조회(원본 데이터 유지)
titanic_df.loc[titanic_df['Age'] >= 65]

PassengerId  Survived  Pclass                                  Name  \
33            34         0       2                 Wheadon, Mr. Edward H   
54            55         0       1        Ostby, Mr. Engelhart Cornelius   
96            97         0       1             Goldschmidt, Mr. George B   
116          117         0       3                  Connors, Mr. Patrick   
280          281         0       3                      Duane, Mr. Frank   
456          457         0       1             Millet, Mr. Francis Davis   
493          494         0       1               Artagaveytia, Mr. Ramon   
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   
672          673         0       2           Mitchell, Mr. Henry Michael   
745          746         0       1          Crosby, Capt. Edward Gifford   
851          852         0       3                   Svensson, Mr. Johan   

      Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
33   male  66.0      0      0  C.A. 24579  10.5000   NaN        S  
54   male  65.0      0      1      113509  61.9792   B30        C  
96   male  71.0      0      0    PC 17754  34.6542    A5        C  
116  male  70.5      0      0      370369   7.7500   NaN        Q  
280  male  65.0      0      0      336439   7.7500   NaN        Q  
456  male  65.0      0      0       13509  26.5500   E38        S  
493  male  71.0      0      0    PC 17609  49.5042   NaN        C  
630  male  80.0      0      0       27042  30.0000   A23        S  
672  male  70.0      0      0  C.A. 24580  10.5000   NaN        S  
745  male  70.0      1      1   WE/P 5735  71.0000   B22        S  
851  male  74.0      0      0      347060   7.7750   NaN        S

In [70]:
# 65세 이상인 승객의 나이(Age)와 이름(Name) 조회
titanic_df.loc[titanic_df['Age'] >= 65, ['Age', 'Name']]

Age                                  Name
33   66.0                 Wheadon, Mr. Edward H
54   65.0        Ostby, Mr. Engelhart Cornelius
96   71.0             Goldschmidt, Mr. George B
116  70.5                  Connors, Mr. Patrick
280  65.0                      Duane, Mr. Frank
456  65.0             Millet, Mr. Francis Davis
493  71.0               Artagaveytia, Mr. Ramon
630  80.0  Barkworth, Mr. Algernon Henry Wilson
672  70.0           Mitchell, Mr. Henry Michael
745  70.0          Crosby, Capt. Edward Gifford
851  74.0                   Svensson, Mr. Johan

In [71]:
# 1등급 선실을 사용한 60세 이상의 여성 승객 데이터 조회
titanic_df.loc[(titanic_df['Age'] >= 60) & (titanic_df['Pclass'] == 1) & (titanic_df['Sex'] == 'female')]

PassengerId  Survived  Pclass  \
275          276         1       1   
366          367         1       1   
829          830         1       1   

                                                 Name     Sex   Age  SibSp  \
275                 Andrews, Miss. Kornelia Theodosia  female  63.0      1   
366  Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  female  60.0      1   
829         Stone, Mrs. George Nelson (Martha Evelyn)  female  62.0      0   

     Parch  Ticket     Fare Cabin Embarked  
275      0   13502  77.9583    D7        S  
366      0  110813  75.2500   D37        C  
829      0  113572  80.0000   B28      NaN

### isin()

In [72]:
# 탑승 항구(Embarked)가 'S', 'C'인 데이터 조회(isin 사용)
titanic_df.loc[titanic_df['Embarked'].isin(['S', 'C'])]

# 두 개 이상 지정할 때는 항상 list 형태

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
884          885         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
884      0   SOTON/OQ 392076   7.0500   NaN        S  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  

[812 rows x 12 columns]

## 데이터 정렬
- sort_values()
- 데이터 오름차순/내림차순 정렬

In [73]:
# 데이터를 이름순으로 정렬한 새로운 데이터프레임 생성
titanic_sorted = titanic_df.sort_values(by='Name')
titanic_sorted.head()

PassengerId  Survived  Pclass                                   Name  \
845          846         0       3                    Abbing, Mr. Anthony   
746          747         0       3            Abbott, Mr. Rossmore Edward   
279          280         1       3       Abbott, Mrs. Stanton (Rosa Hunt)   
308          309         0       2                    Abelson, Mr. Samuel   
874          875         1       2  Abelson, Mrs. Samuel (Hannah Wizosky)   

        Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked  
845    male  42.0      0      0  C.A. 5547   7.55   NaN        S  
746    male  16.0      1      1  C.A. 2673  20.25   NaN        S  
279  female  35.0      1      1  C.A. 2673  20.25   NaN        S  
308    male  30.0      1      0  P/PP 3381  24.00   NaN        C  
874  female  28.0      1      0  P/PP 3381  24.00   NaN        C

In [74]:
# 선실 등급(Pclass)와 이름(Name)을 기준으로 데이터 내림차순 정렬
# 앞선 인덱스의 컬럼을 기준으로 먼저 정렬한 후 다음 컬럼을 기준으로 정렬
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False)
titanic_sorted

PassengerId  Survived  Pclass  \
868          869         0       3   
153          154         0       3   
282          283         0       3   
286          287         1       3   
559          560         1       3   
..           ...       ...     ...   
460          461         1       1   
498          499         0       1   
297          298         0       1   
305          306         1       1   
730          731         1       1   

                                                Name     Sex    Age  SibSp  \
868                      van Melkebeke, Mr. Philemon    male    NaN      0   
153                  van Billiard, Mr. Austin Blyler    male  40.50      0   
282                        de Pelsmaeker, Mr. Alfons    male  16.00      0   
286                          de Mulder, Mr. Theodore    male  30.00      0   
559     de Messemaeker, Mrs. Guillaume Joseph (Emma)  female  36.00      1   
..                                               ...     ...    ...    ...   
460                              Anderson, Mr. Harry    male  48.00      0   
498  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.00      1   
297                     Allison, Miss. Helen Loraine  female   2.00      1   
305                   Allison, Master. Hudson Trevor    male   0.92      1   
730                    Allen, Miss. Elisabeth Walton  female  29.00      0   

     Parch    Ticket      Fare    Cabin Embarked  
868      0    345777    9.5000      NaN        S  
153      2  A/5. 851   14.5000      NaN        S  
282      0    345778    9.5000      NaN        S  
286      0    345774    9.5000      NaN        S  
559      0    345572   17.4000      NaN        S  
..     ...       ...       ...      ...      ...  
460      0     19952   26.5500      E12        S  
498      2    113781  151.5500  C22 C26        S  
297      2    113781  151.5500  C22 C26        S  
305      2    113781  151.5500  C22 C26        S  
730      0     24160  211.3375       B5        S  

[891 rows x 12 columns]

In [75]:
# 선실 등급(Pclass)을 기준으로 내림차순 정렬 후 이름(Name)을 기준으로 오름차순 정렬
titanic_df.sort_values(by=['Pclass', 'Name'], ascending=[False, True])

PassengerId  Survived  Pclass                                    Name  \
845          846         0       3                     Abbing, Mr. Anthony   
746          747         0       3             Abbott, Mr. Rossmore Edward   
279          280         1       3        Abbott, Mrs. Stanton (Rosa Hunt)   
365          366         0       3          Adahl, Mr. Mauritz Nils Martin   
401          402         0       3                         Adams, Mr. John   
..           ...       ...     ...                                     ...   
155          156         0       1             Williams, Mr. Charles Duane   
351          352         0       1  Williams-Lambert, Mr. Fletcher Fellows   
55            56         1       1                       Woolner, Mr. Hugh   
555          556         0       1                      Wright, Mr. George   
325          326         1       1                Young, Miss. Marie Grice   

        Sex   Age  SibSp  Parch     Ticket      Fare Cabin Embarked  
845    male  42.0      0      0  C.A. 5547    7.5500   NaN        S  
746    male  16.0      1      1  C.A. 2673   20.2500   NaN        S  
279  female  35.0      1      1  C.A. 2673   20.2500   NaN        S  
365    male  30.0      0      0     C 7076    7.2500   NaN        S  
401    male  26.0      0      0     341826    8.0500   NaN        S  
..      ...   ...    ...    ...        ...       ...   ...      ...  
155    male  51.0      0      1   PC 17597   61.3792   NaN        C  
351    male   NaN      0      0     113510   35.0000  C128        S  
55     male   NaN      0      0      19947   35.5000   C52        S  
555    male  62.0      0      0     113807   26.5500   NaN        S  
325  female  36.0      0      0   PC 17760  135.6333   C32        C  

[891 rows x 12 columns]

## Aggregation 함수 적용
- min(), max(), sum(), count()

In [76]:
# titanic_df 컬럼별 데이터 수 조회
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [77]:
# Age, Fare 컬럼의 평균 조회
titanic_df[['Age', 'Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

### groupby()

In [79]:
# 선실 등급(Pclass)을 기준으로 그룹 지은 데이터를 새로운 변수에 저장, 데이터 타입 확인
titanic_groupby = titanic_df.groupby(by='Pclass')
print(titanic_groupby)
type(titanic_groupby)

# 데이터프레임이 아닌 groupby 객체
# groupby 데이터에 aggregation 함수를 적용하면 데이터프레임 형태로 변환

pandas.core.groupby.generic.DataFrameGroupBy

In [81]:
# 선실 등급(Pclass)별 데이터 수를 조회
titanic_groupby = titanic_df.groupby(by='Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               216       216   216  216  186    216    216     216   216   
2               184       184   184  184  173    184    184     184   184   
3               491       491   491  491  355    491    491     491   491   

        Cabin  Embarked  
Pclass                   
1         176       214  
2          16       184  
3          12       491

In [82]:
# Passengerid, Survived 컬럼의 선실 등급(Pclass)별 데이터 수 조회
titanic_groupby = titanic_df.groupby(by='Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [83]:
# 여러 개의 aggregaton 함수 동시 적용: list 형태
# Age 컬럼의 선실 등급(Pclass)별 count, max, min, mean, sum 값 조회
titanic_df.groupby(by='Pclass')['Age'].agg(['count', 'max', 'min', 'mean', 'sum'])

count   max   min       mean      sum
Pclass                                       
1         186  80.0  0.92  38.233441  7111.42
2         173  70.0  0.67  29.877630  5168.83
3         355  74.0  0.42  25.140620  8924.92

In [84]:
# 컬럼별로 다른 aggregation 함수 적용: dictionary 형태
agg_format = {'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby(by='Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

## 결측치(NULL) 처리

### isna()
- NaN 여부 확인

In [85]:
titanic_df.isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [86]:
# 컬럼별 결측치 개수 조회
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [87]:
# 전체 데이터의 결측치 개수 조회
titanic_df.isna().sum().sum()

866

### fillna()
- 결측치 대체

In [88]:
# Cabin 컬럼의 NaN 값을 'C000'으로 대체
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  C000        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  C000        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500  C000        S

In [89]:
# Age 컬럼의 NaN 값을 평균 나이로 대체
age_mean = titanic_df['Age'].mean()
titanic_df['Age'] = titanic_df['Age'].fillna(age_mean)
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  C000        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  C000        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500  C000        S

In [90]:
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       2
dtype: int64

### dropna()
- 결측치가 포함된 행 제거

In [91]:
# 결측치가 포함된 데이터 삭제
titanic_df = titanic_df.dropna()
titanic_df.shape

# 전체 데이터 개수: 891 -> 889
# Embarked 컬럼이 NaN 값인 두 개의 데이터 삭제

(889, 12)

In [92]:
# Embarked 컬럼의 값이 NaN인 데이터 삭제
titanic_df.dropna(subset=['Embarked'])
titanic_df.shape

(889, 12)

## apply(lambda: )
- 일괄 가공이나 복잡한 데이터 가공이 필요한 경우 apply lambda 사용
- 복잡한 함수를 생성하여 데이터에 일괄 적용 가능

In [93]:
# 각 이름(Name)의 길이를 구하여 Name_len 컬럼에 저장
titanic_df['Name_len'] = titanic_df['Name'].apply(lambda x: len(x))
titanic_df[['Name', 'Name_len']]

Name  Name_len
0                              Braund, Mr. Owen Harris        23
1    Cumings, Mrs. John Bradley (Florence Briggs Th...        51
2                               Heikkinen, Miss. Laina        22
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)        44
4                             Allen, Mr. William Henry        24
..                                                 ...       ...
886                              Montvila, Rev. Juozas        21
887                       Graham, Miss. Margaret Edith        28
888           Johnston, Miss. Catherine Helen "Carrie"        40
889                              Behr, Mr. Karl Howell        21
890                                Dooley, Mr. Patrick        19

[889 rows x 2 columns]

In [96]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성

def get_category(age) :

    if age <= 5:
      cat = 'Baby'
    elif age <= 12:
      cat = 'Child'
    elif age <= 18:
      cat = 'Teenager'
    elif age <= 25:
      cat = 'Student'
    else:
      cat = 'Adult'
        
    return cat

In [97]:
# Age 컬럼에 get_category 함수를 적용한 Age_cat 컬럼 생성
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x: get_category(x))
titanic_df[['Age', 'Age_cat']]

Age  Age_cat
0    22.000000  Student
1    38.000000    Adult
2    26.000000    Adult
3    35.000000    Adult
4    35.000000    Adult
..         ...      ...
886  27.000000    Adult
887  19.000000  Student
888  29.699118    Adult
889  26.000000    Adult
890  32.000000    Adult

[889 rows x 2 columns]

## 피벗테이블(pivot table)
- MS EXCEL의 피벗테이블과 유사한 기능
- DataFrame.pivot_table(index=row, columns=column, values=value, aggfunc=aggregation_function)


In [98]:
import seaborn as sns
import pandas as pd

# tips: seaborn 라이브러리의 toy dataset
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

### 컬럼 정보
- total_bill: 총 청구 금액
- tip: 팁
- sex: 성별
- smoker: 흡연 여부
- day: 요일
- time: 식사 시간
- size: 식사 인원

In [99]:
# day 컬럼의 유니크한 값을 조회
tips['day'].unique()

['Sun', 'Sat', 'Thur', 'Fri']
Categories (4, object): ['Thur', 'Fri', 'Sat', 'Sun']

In [100]:
# 모든 수치형 데이터에 대한 피벗테이블 생성
tips.pivot_table(index='time', columns='day')
# pd.pivot_table(tips, index='time', columns='day')

# 집계 함수(aggfunc) 기본값: mean

size                                     tip                      \
day         Thur       Fri       Sat       Sun      Thur       Fri       Sat   
time                                                                           
Lunch   2.459016  2.000000       NaN       NaN  2.767705  2.382857       NaN   
Dinner  2.000000  2.166667  2.517241  2.842105  3.000000  2.940000  2.993103   

                 total_bill                               
day          Sun       Thur        Fri        Sat    Sun  
time                                                      
Lunch        NaN  17.664754  12.845714        NaN    NaN  
Dinner  3.255132  18.780000  19.663333  20.441379  21.41

In [101]:
# total_bill 값에 대한 피벗테이블(aggfunc='mean')
tips.pivot_table(index='time', columns='day', values='total_bill')

day          Thur        Fri        Sat    Sun
time                                          
Lunch   17.664754  12.845714        NaN    NaN
Dinner  18.780000  19.663333  20.441379  21.41

In [102]:
# total_bill 값에 대한 피벗테이블(aggfunc='count')
tips.pivot_table(index='time', columns='day', values='total_bill', aggfunc='count')

day     Thur  Fri  Sat  Sun
time                       
Lunch     61    7    0    0
Dinner     1   12   87   76

In [103]:
# 여러 개의 index, values, aggfunc 사용(list 활용)
pv_df = tips.pivot_table(index=['time', 'sex'], columns='smoker',
                         values=['tip', 'total_bill'], aggfunc=['mean', 'max'])
pv_df

mean                                   max        \
                    tip           total_bill              tip         
smoker              Yes        No        Yes         No   Yes    No   
time   sex                                                            
Lunch  Male    2.790769  2.941500  17.374615  18.486500   5.0  6.70   
       Female  2.891000  2.459600  17.431000  15.902400   5.0  5.17   
Dinner Male    3.123191  3.158052  23.642553  20.130130  10.0  9.00   
       Female  2.949130  3.044138  18.215652  20.004138   6.5  5.20   

                                 
              total_bill         
smoker               Yes     No  
time   sex                       
Lunch  Male        32.68  41.19  
       Female      43.11  34.83  
Dinner Male        50.81  48.33  
       Female      44.30  35.83

In [105]:
# Pclass(row), Sex(column) 기준으로 Age 값의 평균에 대한 피벗테이블 생성(df1)
df1 = titanic_df.pivot_table(index='Pclass', columns='Sex', values='Age')
df1

Sex        female       male
Pclass                      
1       33.796653  39.287717
2       28.748661  30.653908
3       24.068493  27.372153

In [106]:
# Pclass, Sex 기준으로 생존여부의 mean, sum에 대한 피벗테이블 생성(df2)
df2 = titanic_df.pivot_table(index='Pclass', columns='Sex',
                             values='Survived', aggfunc=['mean', 'sum'])
df2

mean              sum     
Sex       female      male female male
Pclass                                
1       0.967391  0.368852     89   45
2       0.921053  0.157407     70   17
3       0.500000  0.135447     72   47

In [107]:
# 생존여부와 요금(Fare)에 대한 피벗테이블
df2 = titanic_df.pivot_table(index='Pclass', columns='Sex',
                             values=['Survived', 'Fare'],
                             aggfunc=['mean', 'sum'])
df2

mean                                       sum             \
              Fare             Survived                 Fare              
Sex         female       male    female      male     female       male   
Pclass                                                                    
1       106.693750  67.226127  0.967391  0.368852  9815.8250  8201.5875   
2        21.970121  19.741782  0.921053  0.157407  1669.7292  2132.1125   
3        16.118810  12.661633  0.500000  0.135447  2321.1086  4393.5865   

                      
       Survived       
Sex      female male  
Pclass                
1            89   45  
2            70   17  
3            72   47

## 데이터 합치기

### 국내 아이돌 브랜드 평판 데이터
- 한국기업평판연구소 (http://brikorea.com)



### 웹 상의 표데이터 가져오기

In [108]:
import pandas as pd
import numpy as np

In [109]:
url = 'http://brikorea.com/rk/singer2203'

In [110]:
# URL 페이지의 표 데이터 크롤링
table = pd.read_html(url)

In [111]:
# table 길이 출력
print(type(table))
len(table)

# 해당 링크에 두 개의 테이블 존재
# 필요한 데이터는 두 개 중 첫 번째 테이블

<class 'list'>


2

In [112]:
# 첫 번째 테이블 추출(df_table)
df_table = table[0]
df_table.head()

순위    브랜드  Link  브랜드평판지수     참여지수    미디어지수     소통지수   커뮤니티지수    추천지수  추천참여
0   1  방탄소년단   NaN  9536463  1812642  2646560  2756496  2320764     393  추천하기
1   2    아이유   NaN  7196000  1060792  1874688  2121895  2138624     222  추천하기
2   3    임영웅   NaN  6975033  2054859  2522122  1242074  1155978  102084  추천하기
3   4    박재범   NaN  5018356   649325  1544384  1009679  1814968      88  추천하기
4   5    이찬원   NaN  4592393  1575878  1084591   975997   955927   48031  추천하기

### 데이터 합치기 실습 데이터 준비
- concat: df_50, df_100
- merge: df_left, df_right

In [113]:
# df_table 상위 50개 데이터를 df_50에 저장
df_50 = df_table.loc[0:49]
# df_50 = df_table[:50]

df_50.shape

(50, 10)

In [114]:
# df_table 하위 50개 데이터를 df_100에 저장
df_100 = df_table.loc[50:99]
df_100.head()

# 현재 인덱스 50부터 시작

순위  브랜드  Link  브랜드평판지수    참여지수   미디어지수    소통지수  커뮤니티지수  추천지수  추천참여
50  51   라붐   NaN   647488  124411  271408  126615  125055     8  추천하기
51  52   창모   NaN   645366  140664  235505  128108  141089     6  추천하기
52  53   규현   NaN   629700  113461  262071  131366  122802     5  추천하기
53  54  케이윌   NaN   618334  130822  224692  130260  132561     3  추천하기
54  55  고은성   NaN   604757  133542  217680  126549  126986  2223  추천하기

In [115]:
# 인덱스 재설정(0부터 시작)
df_100 = df_100.reset_index(drop=True) # 기존 인덱스 drop
df_100.head()

순위  브랜드  Link  브랜드평판지수    참여지수   미디어지수    소통지수  커뮤니티지수  추천지수  추천참여
0  51   라붐   NaN   647488  124411  271408  126615  125055     8  추천하기
1  52   창모   NaN   645366  140664  235505  128108  141089     6  추천하기
2  53   규현   NaN   629700  113461  262071  131366  122802     5  추천하기
3  54  케이윌   NaN   618334  130822  224692  130260  132561     3  추천하기
4  55  고은성   NaN   604757  133542  217680  126549  126986  2223  추천하기

In [116]:
# df_table 컬럼명 조회
df_table.columns

Index(['순위', '브랜드', 'Link', '브랜드평판지수', '참여지수', '미디어지수', '소통지수', '커뮤니티지수',
       '추천지수', '추천참여'],
      dtype='object')

In [117]:
# '브랜드', '브랜드평판지수', '미디어지수' 컬럼을 df_left에 저장
df_left = df_table.loc[:, ['브랜드', '브랜드평판지수', '미디어지수']]
df_left.head()

브랜드  브랜드평판지수    미디어지수
0  방탄소년단  9536463  2646560
1    아이유  7196000  1874688
2    임영웅  6975033  2522122
3    박재범  5018356  1544384
4    이찬원  4592393  1084591

In [118]:
# '브랜드', '소통지수', '커뮤니티지수' 컬럼을 df_right에 저장
df_right = df_table.loc[:, ['브랜드', '소통지수', '커뮤니티지수']]
df_right.head()

브랜드     소통지수   커뮤니티지수
0  방탄소년단  2756496  2320764
1    아이유  2121895  2138624
2    임영웅  1242074  1155978
3    박재범  1009679  1814968
4    이찬원   975997   955927

### concat

In [119]:
# df_50, df_100 row 기준 concat
pd.concat([df_50, df_100])

# 병합 시 기존 인덱스 그대로 사용(default)

순위    브랜드  Link  브랜드평판지수     참여지수    미디어지수     소통지수   커뮤니티지수    추천지수  \
0     1  방탄소년단   NaN  9536463  1812642  2646560  2756496  2320764     393   
1     2    아이유   NaN  7196000  1060792  1874688  2121895  2138624     222   
2     3    임영웅   NaN  6975033  2054859  2522122  1242074  1155978  102084   
3     4    박재범   NaN  5018356   649325  1544384  1009679  1814968      88   
4     5    이찬원   NaN  4592393  1575878  1084591   975997   955927   48031   
..  ...    ...   ...      ...      ...      ...      ...      ...     ...   
45   96    황치열   NaN   239715    42500   103011    47547    46658    1009   
46   97   10CM   NaN   239469    48952    38556    69324    82637       4   
47   98  NS 윤지   NaN   239338    48898    93121    49083    48237       4   
48   99     탑현   NaN   238527    56744    85587    47827    48369       4   
49  100    한동근   NaN   211285    46136    74783    43125    47242       4   

    추천참여  
0   추천하기  
1   추천하기  
2   추천하기  
3   추천하기  
4   추천하기  
..   ...  
45  추천하기  
46  추천하기  
47  추천하기  
48  추천하기  
49  추천하기  

[100 rows x 10 columns]

#### 병합된 데이터의 인덱스 재설정
- ignore_index=True
 - 데이터 병합과 동시에 인덱스 재설정
- reset_index(drop=True)
 - 병합된 데이터의 인덱스 재설정
 - drop=True : 인덱스 재설정 시 기존 인덱스 삭제

In [120]:
# 데이터 병합과 동시에 인덱스 재설정(ignore_index=True)
pd.concat([df_50, df_100], ignore_index=True)

순위    브랜드  Link  브랜드평판지수     참여지수    미디어지수     소통지수   커뮤니티지수    추천지수  \
0     1  방탄소년단   NaN  9536463  1812642  2646560  2756496  2320764     393   
1     2    아이유   NaN  7196000  1060792  1874688  2121895  2138624     222   
2     3    임영웅   NaN  6975033  2054859  2522122  1242074  1155978  102084   
3     4    박재범   NaN  5018356   649325  1544384  1009679  1814968      88   
4     5    이찬원   NaN  4592393  1575878  1084591   975997   955927   48031   
..  ...    ...   ...      ...      ...      ...      ...      ...     ...   
95   96    황치열   NaN   239715    42500   103011    47547    46658    1009   
96   97   10CM   NaN   239469    48952    38556    69324    82637       4   
97   98  NS 윤지   NaN   239338    48898    93121    49083    48237       4   
98   99     탑현   NaN   238527    56744    85587    47827    48369       4   
99  100    한동근   NaN   211285    46136    74783    43125    47242       4   

    추천참여  
0   추천하기  
1   추천하기  
2   추천하기  
3   추천하기  
4   추천하기  
..   ...  
95  추천하기  
96  추천하기  
97  추천하기  
98  추천하기  
99  추천하기  

[100 rows x 10 columns]

In [121]:
# df_left, df_right 컬럼 기준 concat(axis=1)
pd.concat([df_left, df_right], axis=1)

브랜드  브랜드평판지수    미디어지수    브랜드     소통지수   커뮤니티지수
0   방탄소년단  9536463  2646560  방탄소년단  2756496  2320764
1     아이유  7196000  1874688    아이유  2121895  2138624
2     임영웅  6975033  2522122    임영웅  1242074  1155978
3     박재범  5018356  1544384    박재범  1009679  1814968
4     이찬원  4592393  1084591    이찬원   975997   955927
..    ...      ...      ...    ...      ...      ...
95    황치열   239715   103011    황치열    47547    46658
96   10CM   239469    38556   10CM    69324    82637
97  NS 윤지   239338    93121  NS 윤지    49083    48237
98     탑현   238527    85587     탑현    47827    48369
99    한동근   211285    74783    한동근    43125    47242

[100 rows x 6 columns]

In [122]:
# df_right의 4, 11번 행을 삭제하여 df_temp에 저장
df_temp = df_right.drop([4, 11])
df_temp.head(20)

# 0부터 21까지의 인덱스(4, 11 제외)

브랜드     소통지수   커뮤니티지수
0     방탄소년단  2756496  2320764
1       아이유  2121895  2138624
2       임영웅  1242074  1155978
3       박재범  1009679  1814968
5        태연  1338896  1270599
6      강다니엘   934054   949383
7   (여자)아이들  1051722   991618
8      이솔로몬   694382   699224
9        영탁   640498   526258
10     레드벨벳   788054   720160
12      이승윤   519406   532833
13      정동원   582594   503329
14      이무진   497634   503923
15     스테이씨   466320   501922
16      세븐틴   496991   505282
17      이승기   411415   427903
18     트와이스   375564   397696
19      장민호   338026   355713
20    하이라이트   433480   339495
21     오마이걸   291951   344611

In [123]:
# df_left, df_temp 컬럼 기준 concat 후 데이터 확인(인덱스 재설정 전)
pd.concat([df_left, df_temp], axis=1).head(20)

# 4, 11번 행이 삭제된 df_temp의 빈 데이터 NaN 처리

브랜드  브랜드평판지수    미디어지수      브랜드       소통지수     커뮤니티지수
0     방탄소년단  9536463  2646560    방탄소년단  2756496.0  2320764.0
1       아이유  7196000  1874688      아이유  2121895.0  2138624.0
2       임영웅  6975033  2522122      임영웅  1242074.0  1155978.0
3       박재범  5018356  1544384      박재범  1009679.0  1814968.0
4       이찬원  4592393  1084591      NaN        NaN        NaN
5        태연  4133803   975184       태연  1338896.0  1270599.0
6      강다니엘  3514389  1042596     강다니엘   934054.0   949383.0
7   (여자)아이들  3206271   828290  (여자)아이들  1051722.0   991618.0
8      이솔로몬  2931481   803522     이솔로몬   694382.0   699224.0
9        영탁  2767955   702477       영탁   640498.0   526258.0
10     레드벨벳  2755590   905079     레드벨벳   788054.0   720160.0
11     블랙핑크  2699549   970185      NaN        NaN        NaN
12      이승윤  2616814  1007249      이승윤   519406.0   532833.0
13      정동원  2587351   893250      정동원   582594.0   503329.0
14      이무진  2521390   988754      이무진   497634.0   503923.0
15     스테이씨  2430642   916966     스테이씨   466320.0   501922.0
16      세븐틴  2419128   916122      세븐틴   496991.0   505282.0
17      이승기  2033473   798272      이승기   411415.0   427903.0
18     트와이스  1794932   702340     트와이스   375564.0   397696.0
19      장민호  1750277   715061      장민호   338026.0   355713.0

In [124]:
# df_temp 인덱스 재설정
df_temp = df_temp.reset_index(drop=True)
df_temp.head(20)

브랜드     소통지수   커뮤니티지수
0     방탄소년단  2756496  2320764
1       아이유  2121895  2138624
2       임영웅  1242074  1155978
3       박재범  1009679  1814968
4        태연  1338896  1270599
5      강다니엘   934054   949383
6   (여자)아이들  1051722   991618
7      이솔로몬   694382   699224
8        영탁   640498   526258
9      레드벨벳   788054   720160
10      이승윤   519406   532833
11      정동원   582594   503329
12      이무진   497634   503923
13     스테이씨   466320   501922
14      세븐틴   496991   505282
15      이승기   411415   427903
16     트와이스   375564   397696
17      장민호   338026   355713
18    하이라이트   433480   339495
19     오마이걸   291951   344611

In [125]:
# df_left, df_temp 컬럼 기준 concat 후 데이터 확인(인덱스 재설정 후)
pd.concat([df_left, df_temp], axis=1).head(20)

# concat은 인덱스 기준으로 병합되므로 서로 다른 데이터가 합쳐진 것 확인

브랜드  브랜드평판지수    미디어지수      브랜드       소통지수     커뮤니티지수
0     방탄소년단  9536463  2646560    방탄소년단  2756496.0  2320764.0
1       아이유  7196000  1874688      아이유  2121895.0  2138624.0
2       임영웅  6975033  2522122      임영웅  1242074.0  1155978.0
3       박재범  5018356  1544384      박재범  1009679.0  1814968.0
4       이찬원  4592393  1084591       태연  1338896.0  1270599.0
5        태연  4133803   975184     강다니엘   934054.0   949383.0
6      강다니엘  3514389  1042596  (여자)아이들  1051722.0   991618.0
7   (여자)아이들  3206271   828290     이솔로몬   694382.0   699224.0
8      이솔로몬  2931481   803522       영탁   640498.0   526258.0
9        영탁  2767955   702477     레드벨벳   788054.0   720160.0
10     레드벨벳  2755590   905079      이승윤   519406.0   532833.0
11     블랙핑크  2699549   970185      정동원   582594.0   503329.0
12      이승윤  2616814  1007249      이무진   497634.0   503923.0
13      정동원  2587351   893250     스테이씨   466320.0   501922.0
14      이무진  2521390   988754      세븐틴   496991.0   505282.0
15     스테이씨  2430642   916966      이승기   411415.0   427903.0
16      세븐틴  2419128   916122     트와이스   375564.0   397696.0
17      이승기  2033473   798272      장민호   338026.0   355713.0
18     트와이스  1794932   702340    하이라이트   433480.0   339495.0
19      장민호  1750277   715061     오마이걸   291951.0   344611.0

### merge
- how : left, right, inner(default), outer

In [126]:
# 브랜드를 기준으로 right merge
pd.merge(df_left, df_temp, on='브랜드', how='right')

# concat은 행 인덱스 기준 병합, merge는 기준 컬럼 지정

브랜드  브랜드평판지수    미디어지수     소통지수   커뮤니티지수
0   방탄소년단  9536463  2646560  2756496  2320764
1     아이유  7196000  1874688  2121895  2138624
2     임영웅  6975033  2522122  1242074  1155978
3     박재범  5018356  1544384  1009679  1814968
4      태연  4133803   975184  1338896  1270599
..    ...      ...      ...      ...      ...
93    황치열   239715   103011    47547    46658
94   10CM   239469    38556    69324    82637
95  NS 윤지   239338    93121    49083    48237
96     탑현   238527    85587    47827    48369
97    한동근   211285    74783    43125    47242

[98 rows x 5 columns]

In [127]:
# 브랜드를 기준으로 inner merge(교집합)
pd.merge(df_left, df_temp, on='브랜드', how='inner')

브랜드  브랜드평판지수    미디어지수     소통지수   커뮤니티지수
0   방탄소년단  9536463  2646560  2756496  2320764
1     아이유  7196000  1874688  2121895  2138624
2     임영웅  6975033  2522122  1242074  1155978
3     박재범  5018356  1544384  1009679  1814968
4      태연  4133803   975184  1338896  1270599
..    ...      ...      ...      ...      ...
93    황치열   239715   103011    47547    46658
94   10CM   239469    38556    69324    82637
95  NS 윤지   239338    93121    49083    48237
96     탑현   238527    85587    47827    48369
97    한동근   211285    74783    43125    47242

[98 rows x 5 columns]

#### 컬럼명이 다른 동일한 데이터의 merge
- left_on, right_on

In [128]:
df_temp.head()

브랜드     소통지수   커뮤니티지수
0  방탄소년단  2756496  2320764
1    아이유  2121895  2138624
2    임영웅  1242074  1155978
3    박재범  1009679  1814968
4     태연  1338896  1270599

In [129]:
# df_temp 컬럼명 수정(브랜드 -> 이름)
df_temp.columns = ['이름', '소통지수', '커뮤니티지수']
df_temp.head()

이름     소통지수   커뮤니티지수
0  방탄소년단  2756496  2320764
1    아이유  2121895  2138624
2    임영웅  1242074  1155978
3    박재범  1009679  1814968
4     태연  1338896  1270599

In [130]:
# left_on='브랜드', right_on='이름' outer merge
merge_df = pd.merge(df_left, df_temp, left_on='브랜드', right_on='이름', how='outer')
merge_df

브랜드  브랜드평판지수    미디어지수     이름       소통지수     커뮤니티지수
0   방탄소년단  9536463  2646560  방탄소년단  2756496.0  2320764.0
1     아이유  7196000  1874688    아이유  2121895.0  2138624.0
2     임영웅  6975033  2522122    임영웅  1242074.0  1155978.0
3     박재범  5018356  1544384    박재범  1009679.0  1814968.0
4     이찬원  4592393  1084591    NaN        NaN        NaN
..    ...      ...      ...    ...        ...        ...
95    황치열   239715   103011    황치열    47547.0    46658.0
96   10CM   239469    38556   10CM    69324.0    82637.0
97  NS 윤지   239338    93121  NS 윤지    49083.0    48237.0
98     탑현   238527    85587     탑현    47827.0    48369.0
99    한동근   211285    74783    한동근    43125.0    47242.0

[100 rows x 6 columns]

In [131]:
# left_on='브랜드', right_on='이름' inner merge
merge_df = pd.merge(df_left, df_temp, left_on='브랜드', right_on='이름')
merge_df

브랜드  브랜드평판지수    미디어지수     이름     소통지수   커뮤니티지수
0   방탄소년단  9536463  2646560  방탄소년단  2756496  2320764
1     아이유  7196000  1874688    아이유  2121895  2138624
2     임영웅  6975033  2522122    임영웅  1242074  1155978
3     박재범  5018356  1544384    박재범  1009679  1814968
4      태연  4133803   975184     태연  1338896  1270599
..    ...      ...      ...    ...      ...      ...
93    황치열   239715   103011    황치열    47547    46658
94   10CM   239469    38556   10CM    69324    82637
95  NS 윤지   239338    93121  NS 윤지    49083    48237
96     탑현   238527    85587     탑현    47827    48369
97    한동근   211285    74783    한동근    43125    47242

[98 rows x 6 columns]